server/src/main/java/org/opensearch/search/querygroup/tracker/QueryGroupResourceUsageTrackerService.java

```diff
@@ -0,0 +1,123 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.querygroup.tracker;
+
+import org.opensearch.common.inject.Inject;
+import org.opensearch.search.querygroup.QueryGroupLevelResourceUsageView;
+import org.opensearch.search.querygroup.QueryGroupTask;
+import org.opensearch.search.resourcetypes.ResourceType;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskManager;
+import org.opensearch.tasks.TaskResourceTrackingService;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This class tracks requests per QueryGroup
+ */
+// @ExperimentalApi
+public class QueryGroupResourceUsageTrackerService implements QueryGroupUsageTracker, TaskManager.TaskEventListeners {
+
+    public static final List<ResourceType> TRACKED_RESOURCES = List.of(ResourceType.fromName("Memory"), ResourceType.fromName("CPU"));
+
+    private final TaskManager taskManager;
+    private final TaskResourceTrackingService taskResourceTrackingService;
+
+    /**
+     * QueryGroupResourceTrackerService constructor
+     *
+     * @param taskManager Task Manager service for keeping track of currently running tasks on the nodes
+     * @param taskResourceTrackingService Service that helps track resource usage of tasks running on a node.
+     */
+    @Inject
+    public QueryGroupResourceUsageTrackerService(
+        final TaskManager taskManager,
+        final TaskResourceTrackingService taskResourceTrackingService
+    ) {
+        this.taskManager = taskManager;
+        this.taskResourceTrackingService = taskResourceTrackingService;
+    }
+
+    /**
+     * Constructs a map of QueryGroupLevelResourceUsageView instances for each QueryGroup.
+     *
+     * @return Map of QueryGroup views
+     */
+    @Override
+    public Map<String, QueryGroupLevelResourceUsageView> constructQueryGroupLevelUsageViews() {
+        Map<String, QueryGroupLevelResourceUsageView> queryGroupViews = new HashMap<>();
+
+        Map<String, List<Task>> tasksByQueryGroup = getTasksGroupedByQueryGroup();
+        Map<String, Map<ResourceType, Long>> queryGroupResourceUsage = getResourceUsageOfQueryGroups(tasksByQueryGroup);
+
+        for (String queryGroupId : tasksByQueryGroup.keySet()) {
+            QueryGroupLevelResourceUsageView queryGroupLevelResourceUsageView = new QueryGroupLevelResourceUsageView(
+                queryGroupId,
+                queryGroupResourceUsage.get(queryGroupId),
+                tasksByQueryGroup.get(queryGroupId)
+            );
+            queryGroupViews.put(queryGroupId, queryGroupLevelResourceUsageView);
+        }
+        return queryGroupViews;
+    }
+
+    /**
+     * Groups tasks by their associated QueryGroup.
+     *
+     * @return Map of tasks grouped by QueryGroup
+     */
+    private Map<String, List<Task>> getTasksGroupedByQueryGroup() {
```

Comment:
This logic need to change since the `Task` carry `queryGroupId` header not as a field